clients/src/main/java/org/apache/kafka/clients/consumer/internals/ListOffsetsRequestManager.java

```diff
@@ -0,0 +1,346 @@
+        MultiNodeRequest multiNodeRequest = new MultiNodeRequest(timestampsToSearchByNode.size());
+        addMultiNodeRequest(listOffsetsRequestState, multiNodeRequest);
+
+        for (Map.Entry<Node, Map<TopicPartition, ListOffsetsRequestData.ListOffsetsPartition>> entry : timestampsToSearchByNode.entrySet()) {
+            Node node = entry.getKey();
+            ListOffsetsRequest.Builder builder = ListOffsetsRequest.Builder
+                    .forConsumer(requireTimestamps, isolationLevel, false)
+                    .setTargetTimes(ListOffsetsRequest.toListOffsetsTopics(entry.getValue()));
+
+            log.debug("Creating ListOffsetRequest {} for broker {}", builder, node);
+
+            NetworkClientDelegate.UnsentRequest unsentRequest = new NetworkClientDelegate.UnsentRequest(
+                    builder,
+                    Optional.ofNullable(node));
+            unsentRequests.add(unsentRequest);
+            unsentRequest.future().whenComplete((response, error) -> {
+                if (error != null) {
+                    log.error("Sending ListOffsetRequest {} to broker {} failed",
+                            builder,
+                            node,
+                            error);
+                    multiNodeRequest.resultFuture.completeExceptionally(error);
+                } else {
+                    ListOffsetsResponse lor = (ListOffsetsResponse) response.responseBody();
+                    log.trace("Received ListOffsetResponse {} from broker {}", lor, node);
+                    onSendToNodeSuccess(lor, multiNodeRequest);
+                }
+            });
+        }
+        return unsentRequests;
+    }
+
+    private void addMultiNodeRequest(ListOffsetsRequestState listOffsetsRequestState,
+                                     MultiNodeRequest multiNodeRequest) {
+        multiNodeRequest.resultFuture.whenComplete((multiNodeResult, error) -> {
+            // Done sending request to a set of known leaders
+            if (error == null) {
+                listOffsetsRequestState.fetchedOffsets.putAll(multiNodeResult.fetchedOffsets);
+                listOffsetsRequestState.remainingToSearch.keySet().retainAll(multiNodeResult.partitionsToRetry);
+                offsetFetcherUtils.updateSubscriptionState(
+                        multiNodeResult.fetchedOffsets,
+                        isolationLevel);
+
+                if (listOffsetsRequestState.remainingToSearch.size() == 0) {
+                    ListOffsetResult listOffsetResult =
+                            new ListOffsetResult(listOffsetsRequestState.fetchedOffsets,
+                                    listOffsetsRequestState.remainingToSearch.keySet());
+                    listOffsetsRequestState.globalResult.complete(listOffsetResult.offsetAndMetadataMap());
+                }
+            } else {
+                requestsToRetry.add(listOffsetsRequestState);
```

Comment:
Why do we want to retry if the result has been completed exceptionally?

Reply:
Because the failure may be due to a condition that will be fixed on the next metadata update (ex. unknown topic/partition/leader epoch errors), so we want to retry those after the next metadta update. That being said, this part is missing the differentiation between these retriable errors and the non-retriable one. Fixed it and testing it, will be added in following commit

Reply:
I think what confuses me is that, once you complete the future exceptionally, wouldn't the user get the results back? So I don't know why do we want to retry after that.

Reply:
Agree, this mistakenly had a mix of intentions for the retriable and non-retriable errors. Fixed so we keep the requestToRetry when there are retriable errors in the response (future not completed), and keep nothing to retry when there is a failure (covered in tests also with retriable errors vs non-retriable ones)